cpp/calculate-demand.cpp

```diff
@@ -73,45 +59,79 @@ long readUnsatisfiedDemand(std::string targetFilename) {
   return atol(metadataValue);
 }
 
-long calculateUnsatisfiedDemand(std::string targetFilename, std::string demoFilename, std::vector<std::string> facilities, std::vector<float> capacities) {
-
-  // generate target dataset based on demoDataset
-  GDALDataset* demoDataset = openRaster(demoFilename);
-  GDALRasterBand* demoBand = demoDataset->GetRasterBand(1);
-  assert(demoBand->GetRasterDataType() == GDT_Float32);
-  int xBlockSize, yBlockSize;
-  demoBand->GetBlockSize(&xBlockSize, &yBlockSize);
-  GDALDataset* targetDataset = copyRaster(demoDataset, targetFilename, xBlockSize, yBlockSize);
-  GDALRasterBand* targetBand = targetDataset->GetRasterBand(1);
-  closeRaster(demoDataset);
+long calculateUnsatisfiedDemand(std::string targetFilename, std::string demoFilename, float saturation, std::vector<std::string> facilities, std::vector<float> capacities) {
+
+  #ifdef BENCHMARK
+  boost::timer::auto_cpu_timer t(std::cerr, 6, "Total time elapsed: %t sec CPU, %w sec real\n");
+  #endif
+
+  GDALDataset* targetDataset = NULL;
+  int xBlockSize, yBlockSize, targetXSize, targetYSize, targetNXBlocks, targetNYBlocks;
+  int nXValid, nYValid, xOffset, yOffset, dataOffset;
+  float* buffer;
+  float* data;
+  double demoProjection[6];
+  double facilityProjection[6];
+  float nodata;
+  std::string demoProjectionWKT;
+
+  /****************************************************************************
+   * Read demographics raster into memory buffer
+   ****************************************************************************/
+  {
+    #ifdef BENCHMARK
+    boost::timer::auto_cpu_timer t(std::cerr, 6, "Read demographics raster: %t sec CPU, %w sec real\n");
+    #endif
+    GDALDataset* demoDataset = openRaster(demoFilename);
+    GDALRasterBand* demoBand = demoDataset->GetRasterBand(1);
+    assert(demoBand->GetRasterDataType() == GDT_Float32);
+    demoBand->GetBlockSize(&xBlockSize, &yBlockSize);
+    demoDataset->GetGeoTransform(demoProjection);
+
+    targetXSize = demoDataset->GetRasterXSize();
+    targetYSize = demoDataset->GetRasterYSize();
+    targetNXBlocks = (targetXSize + xBlockSize - 1)/xBlockSize;
+    targetNYBlocks = (targetYSize + yBlockSize - 1)/yBlockSize;
+    nodata = demoBand->GetNoDataValue();
+    demoProjectionWKT = demoDataset->GetProjectionRef();
+    data = (float*) CPLMalloc(sizeof(float) * (xBlockSize * yBlockSize) * (targetNXBlocks * targetNYBlocks));
```

Comment:
I understand that [CPLMalloc](http://www.gdal.org/cpl__conv_8h.html#a9a806de98fbddb1337efdb18651aa0f7) raises if it fails.